### PR TITLE
[WIP] Debug API

### PR DESCRIPTION
**Pull request by Sonicadvance1**
This change is [Reviewable](https://reviewable.io/reviews/dolphin-emu/dolphin/3031)

### CODE REVIEW

**Comment by Sonicadvance1** [comment]
The main features of this that I love is that I can disable parts of the core, enable/disable profiling, dump profile information, and dump code blocks.
This is so nice when working under Android, and gives a bit of normality when working under a terrible OS such as Android.

**Comment by Sonicadvance1** [comment]
Some things that still need to be implemented
* Arbitrary memory reading/writing (For both debug interface code viewing, and socketed hacking of games, TPP for example)
* Register reading/writing
* Breakpoints (PC and memory accesses?)
* ...More?

**Comment by degasus** [comment]
"... More?"
Sounds like a good idea to get rid of our built-in debugger. If we're able to get all of this over network, we can just provide an additional GUI. QT will love you for this :D

File: Source/Core/DebuggerTest/Main.cpp [hunk 1]
```diff
@@ -0,0 +1,247 @@
...
+#include "UICommon/DebugInterface.h"
+#include "UICommon/Disassembler.h"
+std::unique_ptr<DebugAPI::DebugHandler> m_client;
+
+// Stupid things
+void Host_NotifyMapLoaded() {}
+void Host_RefreshDSPDebuggerWindow() {}
+void Host_Message(int Id) {}
+
+void* Host_GetRenderHandle() { return NULL; }
```
Comment (lioncash) [inline Source/Core/DebuggerTest/Main.cpp]: nullptr

File: Source/Core/DebuggerTest/Main.cpp [hunk 2]
```diff
@@ -0,0 +1,247 @@
...
+			"JIT Floating Point",
+			"JIT Integer",
+			"JIT Paired",
+			"JIT System Registers",
+		};
+
+		if (disable == "disable")
+			disable_var = true;
+		else
+			disable_var = false;
```
Comment (lioncash) [inline Source/Core/DebuggerTest/Main.cpp]: bool disable_var = (disable == "disable");

File: Source/Core/UICommon/DebugInterface.cpp [hunk 3]
```diff
@@ -0,0 +1,724 @@
...
+	{
+		m_packet_handled.Set();
+	}
+	ClientDebugHandler(std::string ip)
+		: m_ip(ip)
+	{
+		if (enet_initialize() != 0)
+			goto cleanup;
+
+		m_client = enet_host_create(NULL, 1, 2, 57600/8, 14400/8);
```
Comment (lioncash) [inline Source/Core/UICommon/DebugInterface.cpp]: nullptr

File: Source/Core/UICommon/DebugInterface.cpp [hunk 4]
```diff
@@ -0,0 +1,724 @@
...
+	}
+	void WaitForResponse()
+	{
+		m_packet_available.Wait();
+	}
+	void SetPacketHandled()
+	{
+		m_packet_handled.Set();
+	}
+	ClientDebugHandler(std::string ip)
```
Comment (lioncash) [inline Source/Core/UICommon/DebugInterface.cpp]: const reference?

File: Source/Core/UICommon/DebugInterface.cpp [hunk 5]
```diff
@@ -0,0 +1,724 @@
...
+	~ClientDebugHandler()
+	{
+		m_running = false;
+		client_thread.join();
+
+		enet_host_destroy(m_client);
+		enet_deinitialize();
+		m_inited = false;
+	}
+	bool IsAvailable()
```
Comment (lioncash) [inline Source/Core/UICommon/DebugInterface.cpp]: Const function

File: Source/Core/UICommon/DebugInterface.h [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+		virtual void SetCoreFeature(std::string& core_part, bool disabled) = 0;
+		virtual void SetProfile(bool enable) = 0;
+		virtual void GetProfileResults() = 0;
+		virtual void DumpCode(u32 address, std::vector<u32>* ppc, std::vector<u8>* host, u64* host_addr) = 0;
+	};
+
+	class DebugHandlerServerBase
+	{
+	public:
+		virtual bool IsAvailable() = 0;
```
Comment (lioncash) [inline Source/Core/UICommon/DebugInterface.h]: Should probably be const

File: Source/Core/UICommon/DebugInterface.h [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+
+#pragma once
+#include <vector>
+
+namespace DebugAPI
+{
+	class DebugHandler
+	{
+	public:
+		virtual bool IsAvailable() { return true; }
```
Comment (lioncash) [inline Source/Core/UICommon/DebugInterface.h]: Should probably be const

File: Source/Core/UICommon/DebugInterface.cpp [hunk 8]
```diff
@@ -0,0 +1,724 @@
...
+		m_inited = false;
+		m_running = false;
+	}
+	~DebugHandlerServer()
+	{
+		m_inited = false;
+		m_running = false;
+		m_thread.join();
+	}
+	bool IsAvailable()
```
Comment (lioncash) [inline Source/Core/UICommon/DebugInterface.cpp]: const function

**Comment by lioncash** [comment]
Also missing override specifiers.

**Pull request closed without merging**